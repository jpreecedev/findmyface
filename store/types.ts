import { Reducer, AnyAction } from "redux"
import { FormStateMap } from "redux-form"

export interface BasketItem {
  momentId: string
}

export interface PictureItem {
  _id: string
  url: string
  label: string
}

export interface AddBasketItemAction {
  type: string
  payload: BasketItem
}
export interface RemoveBasketItemAction {
  type: string
  payload: BasketItem
}

export interface AddPictureAction {
  type: string
  payload: PictureItem
}

export type BasketActionTypes = AddBasketItemAction | RemoveBasketItemAction

export interface AddPictureAction {
  type: string
  payload: PictureItem
}

export type PicturesActionTypes = AddPictureAction

export type AppState = ReturnType<
  Reducer<
    {
      basket: BasketItem[]
      pictures: PictureItem[]
      form: FormStateMap
    },
    AnyAction
  >
>
