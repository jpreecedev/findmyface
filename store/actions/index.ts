import * as pictures from "./pictures"

export { pictures }
