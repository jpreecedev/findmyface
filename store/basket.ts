import { createSlice } from "@reduxjs/toolkit"

const initialState = []

export const basketSlice = createSlice({
  name: "basket",
  initialState,
  reducers: {
    addPicture(state, action) {
      const exists = state.some(item => item.momentId === action.payload.momentId)
      if (exists) {
        return state
      }

      return [...state, action.payload]
    },
    add(state, action) {
      return state.map(item => {
        if (item.momentId === action.payload.momentId) {
          return {
            ...item,
            addedToBasket: true
          }
        }
        return item
      })
    },
    remove(state, action) {
      return state.map(item => {
        if (item.momentId === action.payload.momentId) {
          return {
            ...item,
            addedToBasket: false
          }
        }
        return item
      })
    },
    clear() {
      return initialState
    }
  }
})

export const { addPicture, add, remove, clear } = basketSlice.actions
