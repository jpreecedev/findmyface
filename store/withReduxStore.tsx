import React from "react"
import { initialiseStore } from "."
import { NextComponentType, NextPageContext } from "next"

const isServer = typeof window === "undefined"
const __NEXT_REDUX_STORE__ = "__NEXT_REDUX_STORE__"

function getOrCreateStore(initialState = {}) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initialiseStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[__NEXT_REDUX_STORE__]) {
    window[__NEXT_REDUX_STORE__] = initialiseStore(initialState)
  }
  return window[__NEXT_REDUX_STORE__]
}

const AppWithRedux = <P extends object>(Component: NextComponentType<P>) =>
  class WithRedux extends React.Component<P> {
    reduxStore: any

    constructor(props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    static async getInitialProps(appContext: any) {
      const reduxStore = getOrCreateStore()

      appContext.ctx.reduxStore = reduxStore

      let appProps = {}
      if (typeof Component.getInitialProps === "function") {
        appProps = await Component.getInitialProps(appContext)
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState()
      }
    }

    render() {
      return <Component {...this.props} reduxStore={this.reduxStore} />
    }
  }

export { AppWithRedux }
