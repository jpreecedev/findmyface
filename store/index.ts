import { combineReducers } from "redux"
import { reducer as formReducer } from "redux-form"
import { configureStore } from "@reduxjs/toolkit"

import { AppState } from "../global"
import { basketSlice } from "./basket"

const loadState = () => {
  try {
    const serializedState = localStorage.getItem("state")
    if (serializedState === null) {
      return {}
    }
    return JSON.parse(serializedState)
  } catch (err) {
    return {}
  }
}
const saveState = (state: AppState) => {
  try {
    const serializedState = JSON.stringify(state)
    localStorage.setItem("state", serializedState)
  } catch {
    // ignore write errors
  }
}

let store

const initialise = (state: AppState = loadState()) => {
  store = configureStore({
    reducer: combineReducers({
      pictures: basketSlice.reducer,
      form: formReducer
    }),
    preloadedState: state
  })

  store.subscribe(() => {
    saveState(<AppState>{
      pictures: store.getState().pictures
    })
  })

  return store
}

export { initialise, store }
