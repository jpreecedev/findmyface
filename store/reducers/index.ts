import { basketReducer } from "./basket"
import { picturesReducer } from "./pictures"

export { basketReducer, picturesReducer }
