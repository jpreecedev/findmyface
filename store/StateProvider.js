import * as React from "react"
import { Provider } from "react-redux"
import { createStore, combineReducers, applyMiddleware, compose } from "redux"
import thunk from "redux-thunk"
import throttle from "lodash.throttle"
import { reducer as formReducer } from "redux-form"

import { basketReducer } from "./reducers"
import { loadState, saveState } from "../utils/localStorage"

const composeEnhancers =
  typeof window === "object" && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
    : compose

const persistedState = loadState()

const reducers = combineReducers({
  basket: basketReducer,
  form: formReducer
})

const store = createStore(
  reducers,
  persistedState,
  composeEnhancers(applyMiddleware(thunk))
)

store.subscribe(
  throttle(() => {
    saveState({
      basket: store.getState().basket
    })
  }, 200)
)

function StateProvider({ children }) {
  return <Provider store={store}>{children}</Provider>
}

export default StateProvider
