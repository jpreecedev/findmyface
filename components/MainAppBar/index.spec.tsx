import React from "react"
import { createStore } from "redux"
import { Provider } from "react-redux"
import { render } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"

import { PictureItem } from "../../global"
import { rootReducer as reducer } from "../../store"
import { PlainMainAppBar } from "."

describe("<MainAppBar /> tests", () => {
  it("should render a default app bar", () => {
    const appbarTestId = "appbar-container"
    const pictures: PictureItem[] = []

    const { getByTestId } = render(<PlainMainAppBar pictures={pictures} />)

    expect(getByTestId(appbarTestId)).toMatchSnapshot()
  })
})
