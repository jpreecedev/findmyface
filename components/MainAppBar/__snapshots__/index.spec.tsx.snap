// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainAppBar /> tests should render a default app bar 1`] = `
<div
  class="makeStyles-grow-1"
  data-testid="appbar-container"
>
  <header
    class="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <h6
        class="MuiTypography-root makeStyles-title-3 MuiTypography-h6 MuiTypography-noWrap"
      >
        Find My Face
      </h6>
      <div
        class="makeStyles-grow-1"
      />
      <div
        class="makeStyles-section-4"
      >
        <button
          aria-label="Basket"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <span
              class="MuiBadge-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17.21 9l-4.38-6.56c-.19-.28-.51-.42-.83-.42-.32 0-.64.14-.83.43L6.79 9H2c-.55 0-1 .45-1 1 0 .09.01.18.04.27l2.54 9.27c.23.84 1 1.46 1.92 1.46h13c.92 0 1.69-.62 1.93-1.46l2.54-9.27L23 10c0-.55-.45-1-1-1h-4.79zM9 9l3-4.4L15 9H9zm3 8c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2z"
                />
              </svg>
              <span
                class="MuiBadge-badge MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorSecondary MuiBadge-invisible"
              >
                0
              </span>
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="account of current user"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
</div>
`;
